mesonbuild/interpreter.py

```diff
@@ -1488,19 +1486,53 @@ def parse_default_options(self, default_options):
                 raise InterpreterException('All default options must be of type key=value.')
             key, value = option.split('=', 1)
             if coredata.is_builtin_option(key):
+                if self.subproject != '':
+                    continue # Only the master project is allowed to set global options.
                 if not self.environment.had_argument_for(key):
                     self.coredata.set_builtin_option(key, value)
                 # If this was set on the command line, do not override.
             else:
+                # Option values set with subproject() default_options override those
+                # set in project() default_options.
+                pref = key + '='
+                for i in self.default_project_options:
+                    if i.startswith(pref):
+                        option = i
+                        break
+                # If we are in a subproject, add the subproject prefix to option
+                # name.
+                if self.subproject != '':
+                    option = self.subproject + ':' + option
                 newoptions = [option] + self.environment.cmd_line_options.projectoptions
                 self.environment.cmd_line_options.projectoptions = newoptions
+        # Add options that are only in default_options.
+        for defopt in self.default_project_options:
+            key, value = defopt.split('=')
+            pref = key + '='
+            was_found = False
+            for i in default_options:
+                if i.startswith(pref):
+                    was_found = True
+                    break
+                if was_found:
+                    break
+            defopt = self.subproject + ':' + defopt
+            newoptions = [defopt] + self.environment.cmd_line_options.projectoptions
+            self.environment.cmd_line_options.projectoptions = newoptions
 
     @stringArgs
     def func_project(self, node, args, kwargs):
+        if self.environment.first_invocation and ('default_options' in kwargs or \
```

Comment:
__[flake8]__
* [E502] the backslash is redundant between brackets

<!-- Comment by [SideCI](https://sideci.com) -->